### DISCUSSION TITLE
Upward connectivities between mesh entities

### DISCUSSION THREAD

**Discussion by ojamond** [libMesh/libmesh#3355] [Q&A]
Hello libMesh users,

I am considering using libMesh as a mesh backend handling distributed meshes with AMR functionnalities in the software I am working on for the french Alternative Energies and Atomic Energy Commission.

I started looking and playing with the classes of this module, but at this time I couldn't find a way to get upward adjacencies between mes entities, i.e. topological connections from lower dimentionnal entities to higher ones (vertex->faces, face->cells, vertex->cells, ...). For example, let's say for a 3D mesh composed of hexahedra, considering a quad face I would like to get the two (or one if the face is located at a booundary) cells adjacent to this face.

Is such a query possible with libMesh?

Many thanks!
Olivier Jamond

**Comment by jwpeterson** [comment]
The short answer is that we don't store the adjacency information in both directions. I think the main reason for this is that, in the `MeshBase`-derived classes, only the cells and vertices are actually stored. The one small exception to this that I can think of is the `Elem::interior_parent()`, which is currently used when building a boundary discretization to allow the boundary elements to know which higher-dimensional element they are a face of.

**Comment by roystgnr** [comment]
If you have an explicitly-constructed quad face, you can call `interior_parent()` ... but in general we don't have explicit faces or edges stored in the mesh; when we need to reference them, a pair of "element, local face/edge number" is the (non-unique!) data used to do so.

For nodes (including vertices), we have a `MeshTools::build_nodes_to_elem_map()` utility that can construct the data you want, but since it's an expense that not all users need we don't build that map by default.

**Comment by ojamond** [comment]
Thanks for your answers!

I tried with the following simple snippet to construct explicitly faces from hex elements, and to call `interior_parent()`, but I always got a null pointer... Is that normal? 

Besides, a face which is not at the boundary of the domain must have two adjacent cells. How `interior_parent()` deals with that?

Thanks for your help!

```c++
#include <iostream>
#include "libmesh/libmesh.h"
#include "libmesh/mesh_generation.h"
#include "libmesh/parallel_mesh.h"
#include "libmesh/elem.h"

int main(int argc, char **argv)
{
  libMesh::LibMeshInit init(argc, argv);

  libMesh::DistributedMesh mesh(init.comm());
  libMesh::MeshTools::Generation::build_cube(mesh, 2, 2, 2);

  for (auto &e : mesh.active_local_element_ptr_range()) {
    for (auto i : e->side_index_range()) {
      auto s = e->build_side_ptr(i, false);
      std::cout << "s->interior_parent(): " << s->interior_parent() << std::endl;
    }
  }

  return 0;
}
```

**Reply by jwpeterson** [reply]
What version of libmesh are you using? From what I can tell, it looks like we call `set_interior_parent` when building sides now, so the pointer should not be `nullptr`. I will also try your test code and check this when I get some time...

> Besides, a face which is not at the boundary of the domain must have two adjacent cells. How interior_parent() deals with that?

The interior_parent is set to the higher-dimensional element that constructs the side. So if you construct a side in the same spatial location from both neighboring higher-dimensional elements, it will not be the "same" in the sense that it will have different interior parents and possibly different orientation, etc.

**Reply by jwpeterson** [reply]
OK, I made one small change to your test code to throw an error if the side's `interior_parent` does not match the current Elem, and this runs for me without errors. So I'm not sure why you are getting a `nullptr`.
```c++
#include <iostream>
#include "libmesh/libmesh.h"
#include "libmesh/mesh_generation.h"
#include "libmesh/parallel_mesh.h"
#include "libmesh/elem.h"

// Test code from libMesh discussion #3355
// https://github.com/libMesh/libmesh/discussions/3355
int main(int argc, char **argv)
{
  libMesh::LibMeshInit init(argc, argv);

  libMesh::DistributedMesh mesh(init.comm());
  libMesh::MeshTools::Generation::build_cube(mesh, 2, 2, 2);

  for (auto &e : mesh.active_local_element_ptr_range()) {
    for (auto i : e->side_index_range()) {
      auto s = e->build_side_ptr(i, false);
      std::cout << "s->interior_parent(): " << s->interior_parent() << std::endl;
      libmesh_error_msg_if(s->interior_parent() != e,
                           "Interior parent of side should be set");
    }
  }

  return 0;
}
```

**Reply by ojamond** [reply]
I use the 1.4.1 version of libmesh... This may explain why I get these null pointers!

I use this version because we manage our dependencies with spack, an this is the more recent version available through spack. I will try to use the last release, which seems to be the 1.7.1 by "overloading" the spack's package.py for libmesh.
